src/utils.rs

```diff
@@ -273,15 +273,43 @@ pub(crate) fn contains_skip(attrs: &[Attribute]) -> bool {
 #[inline]
 pub(crate) fn semicolon_for_expr(context: &RewriteContext<'_>, expr: &ast::Expr) -> bool {
     // Never try to insert semicolons on expressions when we're inside
-    // a macro definition - this can prevent the macro from compiling
+    // a macro definition - this  can prevent the macro from compiling
     // when used in expression position
     if context.is_macro_def {
         return false;
     }
 
     match expr.kind {
         ast::ExprKind::Ret(..) | ast::ExprKind::Continue(..) | ast::ExprKind::Break(..) => {
-            context.config.trailing_semicolon()
+            match context.config.trailing_semicolon() {
+                TrailingSemicolon::Always => true,
+                TrailingSemicolon::Never | TrailingSemicolon::Preserve => false,
+            }
+        }
+        _ => false,
+    }
+}
+
+/// Previously, we used to have `trailing_semicolon = Always` enabled, and due to
+/// a bug between `format_stmt` and `format_expr`, we used to subtract the size of
+/// `;` *TWICE* from the shape. This means that an expr that would fit onto a line
+/// of, e.g. 99 (limit 100) after subtracting one for the semicolon would still be
+/// wrapped.
+///
+/// This function reimplements the old heuristic of double counting the "phantom"
+/// semicolon that should have already been accounted for, to not break existing
+/// formatting with the `trailing_semicolon = Preserve` behavior for Version = One.
+#[inline]
+pub(crate) fn semicolon_for_expr_extra_hacky_double_counted_spacing(
```

Comment:
Do we still need this function if we're version gating the fix and only applying the old behavior for `version=One`?